### PR TITLE
Xpetra&MueLu: ETI and header cleanup

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/xpetra @trilinos/muelu 

## Motivation
- ETI several classes: `IO`, `MatrixMatrix`, `TripleMatrixMultiply`, `Helpers`, `MatrixUtils`, `MatrixFactory`, `ThyraUtils`, `Matrix`, `BlockCrsMatrix`.
- Merge Xpetra libs because the separation into src and sup is broken and only worked due to a lack of ETI. 
- Clean up includes in MueLu::RAPFactory.
- Silence warning in Xpetra::VectorFactory.
- ASAN error in MueLu::Amesos2Smoother
- Bump up tolerances in MueLu Region unit test.
- Tpetra: Change correctness check in Bug8794 unit test

### CODE REVIEW

**Comment by cgcgcg** [comment]
@trilinos/framework What's going on here? Is the AT broken?

**Comment by sebrowne** [comment]
@cgcgcg I checked and the logs are not very helpful.  My money is on a loading error on the machine, and we need to dial down the build parallelism.  I'm going to RETEST again and see what happens.

File: packages/muelu/src/Misc/MueLu_SchurComplementFactory_def.hpp [hunk 1]
```diff
@@ -181,8 +181,8 @@ SchurComplementFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::ComputeSchurC
     if (!A11.is_null()) {
       S = MatrixFactory::BuildCopy(A11);
     } else {
-      S = MatrixFactory::Build(A11->getRowMap(), 10 /*A11->getLocalMaxNumRowEntries()*/);
-      S->fillComplete(A11->getDomainMap(), A11->getRangeMap());
+      TEUCHOS_TEST_FOR_EXCEPTION(true, Exceptions::RuntimeError,
+                                 "MueLu::SchurComplementFactory::Build: Constructing Schur complement for matrix with non-existent diagonal block.");
```
Comment (mayrmt) [inline packages/muelu/src/Misc/MueLu_SchurComplementFactory_def.hpp:185]: @cgcgcg Isn't the case of a zero block in the bottom right place of a 2x2 block matrix the perfect textbook example of a saddle point system suitable for Schur complement approaches?

Comment (cgcgcg) [reply packages/muelu/src/Misc/MueLu_SchurComplementFactory_def.hpp:185]: Not in this case. Further up, this check is false:
https://github.com/trilinos/Trilinos/blob/7252bec780885da9091347ce3ea88e7e14daae5b/packages/muelu/src/Misc/MueLu_SchurComplementFactory_def.hpp#L134
and just above this check is false
https://github.com/trilinos/Trilinos/blob/7252bec780885da9091347ce3ea88e7e14daae5b/packages/muelu/src/Misc/MueLu_SchurComplementFactory_def.hpp#L169

**Comment by sebrowne** [comment]
It's repeatable, so there is likely a real issue here.  I'm traveling today, but I may be able to diagnose it this afternoon (not sure why the specific build error is not showing up on CDash).

**Comment by sebrowne** [comment]
@cgcgcg  looks like shadowing.  Not sure why CDash isn't displaying them....  Here's a sample.  I was able to reproduce them locally:

```
In file included from /fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNames.hpp:49,
                 from /fgs/sebrown/Trilinos/packages/xpetra/sup/Utils/Xpetra_IO_decl.hpp:164,
                 from /fgs/sebrown/Trilinos/build/packages/xpetra/sup/Xpetra_IO.hpp:1,
                 from /fgs/sebrown/Trilinos/packages/xpetra/sup/Utils/Xpetra_IO_def.hpp:48,
                 from /fgs/sebrown/Trilinos/build/packages/xpetra/sup/Utils/ExplicitInstantiation/Xpetra_IO.cpp:55:
/fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNamesScalar.hpp:172:102: note: shadowed declaration is here
 using TpetraVector [[maybe_unused]] = Xpetra::TpetraVector<Scalar, LocalOrdinal, GlobalOrdinal, Node>;
                                                                                                      ^
In file included from /fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNames.hpp:49,
                 from /fgs/sebrown/Trilinos/packages/xpetra/sup/Utils/Xpetra_IO_def.hpp:765,
                 from /fgs/sebrown/Trilinos/build/packages/xpetra/sup/Utils/ExplicitInstantiation/Xpetra_IO.cpp:55:
/fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNamesScalar.hpp:176:102: error: declaration of 'using MapExtractor = class Xpetra::MapExtractor<double, int, long long int, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Serial> >' shadows a previous local [-Werror=shadow]
 using MapExtractor [[maybe_unused]] = Xpetra::MapExtractor<Scalar, LocalOrdinal, GlobalOrdinal, Node>;
                                                                                                      ^
In file included from /fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNames.hpp:49,
                 from /fgs/sebrown/Trilinos/packages/xpetra/sup/Utils/Xpetra_IO_decl.hpp:164,
                 from /fgs/sebrown/Trilinos/build/packages/xpetra/sup/Xpetra_IO.hpp:1,
                 from /fgs/sebrown/Trilinos/packages/xpetra/sup/Utils/Xpetra_IO_def.hpp:48,
                 from /fgs/sebrown/Trilinos/build/packages/xpetra/sup/Utils/ExplicitInstantiation/Xpetra_IO.cpp:55:
/fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNamesScalar.hpp:176:102: note: shadowed declaration is here
 using MapExtractor [[maybe_unused]] = Xpetra::MapExtractor<Scalar, LocalOrdinal, GlobalOrdinal, Node>;
                                                                                                      ^
In file included from /fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNames.hpp:49,
                 from /fgs/sebrown/Trilinos/packages/xpetra/sup/Utils/Xpetra_IO_def.hpp:765,
                 from /fgs/sebrown/Trilinos/build/packages/xpetra/sup/Utils/ExplicitInstantiation/Xpetra_IO.cpp:55:
/fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNamesScalar.hpp:180:116: error: declaration of 'using MapExtractorFactory = class Xpetra::MapExtractorFactory<double, int, long long int, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Serial> >' shadows a previous local [-Werror=shadow]
 using MapExtractorFactory [[maybe_unused]] = Xpetra::MapExtractorFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>;
                                                                                                                    ^
In file included from /fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNames.hpp:49,
                 from /fgs/sebrown/Trilinos/packages/xpetra/sup/Utils/Xpetra_IO_decl.hpp:164,
                 from /fgs/sebrown/Trilinos/build/packages/xpetra/sup/Xpetra_IO.hpp:1,
                 from /fgs/sebrown/Trilinos/packages/xpetra/sup/Utils/Xpetra_IO_def.hpp:48,
                 from /fgs/sebrown/Trilinos/build/packages/xpetra/sup/Utils/ExplicitInstantiation/Xpetra_IO.cpp:55:
/fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNamesScalar.hpp:180:116: note: shadowed declaration is here
 using MapExtractorFactory [[maybe_unused]] = Xpetra::MapExtractorFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>;
                                                                                                                    ^
In file included from /fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNames.hpp:49,
                 from /fgs/sebrown/Trilinos/packages/xpetra/sup/Utils/Xpetra_IO_def.hpp:765,
                 from /fgs/sebrown/Trilinos/build/packages/xpetra/sup/Utils/ExplicitInstantiation/Xpetra_IO.cpp:55:
/fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNamesScalar.hpp:186:35: error: declaration of 'using SC = double' shadows a previous local [-Werror=shadow]
 using SC [[maybe_unused]] = Scalar;
                                   ^
In file included from /fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNames.hpp:49,
                 from /fgs/sebrown/Trilinos/packages/xpetra/sup/Utils/Xpetra_IO_decl.hpp:164,
                 from /fgs/sebrown/Trilinos/build/packages/xpetra/sup/Xpetra_IO.hpp:1,
                 from /fgs/sebrown/Trilinos/packages/xpetra/sup/Utils/Xpetra_IO_def.hpp:48,
                 from /fgs/sebrown/Trilinos/build/packages/xpetra/sup/Utils/ExplicitInstantiation/Xpetra_IO.cpp:55:
/fgs/sebrown/Trilinos/packages/xpetra/src/Headers/Xpetra_UseShortNamesScalar.hpp:186:35: note: shadowed declaration is here
 using SC [[maybe_unused]] = Scalar;
```

**Comment by cgcgcg** [comment]
@sebrowne These are warning though. They don't get elevated to errors, do they?

**Review by etphipp — DISMISSED** [review]
Stokhos changes are fine.

**Review by jhux2 — APPROVED** [review]
Lgtm (reviewed twice with @cgcgcg)

**Pull request merged**